lib/Transforms/LowerPolynomialEval/Patterns.cpp

```diff
@@ -0,0 +1,201 @@
+
+#define DEBUG_TYPE "lower-polynomial-eval"
+
+namespace mlir {
+namespace heir {
+
+using polynomial::EvalOp;
+using polynomial::TypedFloatPolynomialAttr;
+
+TypedAttr getScalarOrDenseAttr(Type tensorOrScalarType, APFloat value) {
+  return TypeSwitch<Type, TypedAttr>(tensorOrScalarType)
+      .Case<FloatType>(
+          [&](FloatType type) { return FloatAttr::get(type, value); })
+      .Case<ShapedType>(
+          [&](ShapedType type) { return DenseElementsAttr::get(type, value); })
+      .Default([](Type) { return nullptr; });
+}
+
+LogicalResult LowerViaHorner::matchAndRewrite(EvalOp op,
+                                              PatternRewriter& rewriter) const {
+  Type evaluatedType = op.getValue().getType();
+  ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+  b.setInsertionPoint(op);
+
+  LLVM_DEBUG(llvm::dbgs() << "evaluatedType: " << evaluatedType << "\n");
+
+  auto attr =
+      dyn_cast<polynomial::TypedFloatPolynomialAttr>(op.getPolynomialAttr());
+  if (!attr) return failure();
+
+  auto terms = attr.getValue().getPolynomial().getTerms();
+  int64_t maxDegree = terms.back().getExponent().getSExtValue();
+  const int degreeThreshold = 5;
+  if (!shouldForce() && maxDegree >= degreeThreshold) return failure();
+
+  auto monomialMap = attr.getValue().getPolynomial().getCoeffMap();
+  DenseMap<int64_t, TypedAttr> attributeMap;
+  for (auto& [key, monomial] : monomialMap) {
+    attributeMap[key] =
+        getScalarOrDenseAttr(evaluatedType, monomial.getCoefficient());
+  }
+
+  // Start with the coefficient of the highest degree term
+  Value result =
+      b.create<arith::ConstantOp>(evaluatedType, attributeMap[maxDegree]);
+
+  // Apply Horner's method, accounting for possible missing terms
+  auto x = op.getOperand();
+  for (int64_t i = maxDegree - 1; i >= 0; i--) {
+    // Multiply by x
+    result = b.create<arith::MulFOp>(result, x);
+
+    // Add coefficient if this term exists, otherwise continue
+    if (attributeMap.find(i) != attributeMap.end()) {
+      auto coeffConst =
+          b.create<arith::ConstantOp>(evaluatedType, attributeMap[i]);
+      result = b.create<arith::AddFOp>(result, coeffConst);
+    }
+  }
+
+  rewriter.replaceOp(op, result);
+  return success();
+}
+
+LogicalResult LowerViaPatersonStockmeyerMonomial::matchAndRewrite(
+    EvalOp op, PatternRewriter& rewriter) const {
+  Type evaluatedType = op.getValue().getType();
+  ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+  b.setInsertionPoint(op);
+
+  auto attr =
+      dyn_cast<polynomial::TypedFloatPolynomialAttr>(op.getPolynomialAttr());
+  if (!attr) return failure();
+
+  auto terms = attr.getValue().getPolynomial().getTerms();
+
+  if (terms.empty()) {
+    // An empty polynomial evaluates to zero.
+    Value zeroConst = b.create<arith::ConstantOp>(evaluatedType,
+                                                  b.getZeroAttr(evaluatedType));
+    rewriter.replaceOp(op, zeroConst);
+    return success();
+  }
```

Comment:
this is covered by the verifier, right, so shouldn't ever hit? (why isn't an empty poly eval not a folder?)

Reply:
Correct, removed!